lib/automation.go

```diff
@@ -126,3 +147,183 @@ func (automationImpl) Apply(scope scope, p *ApplyParams) (*ApplyResult, error) {
 	res.RunID = runID
 	return res, nil
 }
+
+// Deploy adds or updates a Dataset, creates or updates an associated Workflow, and, if deployParams.Apply is true, immediately runs the Workflow
+func (automationImpl) Deploy(scope scope, p *DeployParams) error {
+	log.Debugw("deploy", "dataset name", p.Dataset.Name, "peername", p.Dataset.Peername, "workflow id", p.Workflow.ID)
+	if p.Workflow.ID != "" {
+		wf, err := scope.Automation().GetWorkflow(p.Workflow.ID)
+		if err != nil {
+			return nil
+		}
+		if p.Workflow.DatasetID != wf.DatasetID {
+			return fmt.Errorf("deploy: given workflow and workflow on record have different DatasetIDs")
+		}
+		if p.Workflow.OwnerID != wf.OwnerID {
+			return fmt.Errorf("deploy: given workflow and workflow on record have different OwnerIDs")
+		}
+	}
+	go deploy(scope, p)
+	return nil
+}
+
+func deploy(scope scope, p *DeployParams) {
+	vi := dsref.ConvertDatasetToVersionInfo(p.Dataset)
+	ref := vi.SimpleRef().String()
+	deployPayload := event.DeployEvent{
+		Ref:        ref,
+		DatasetID:  p.Dataset.ID,
+		WorkflowID: p.Workflow.ID.String(),
+	}
+	log.Debugw("deploy started", "payload", deployPayload)
+	if scope.Bus() != nil {
+		scope.Bus().PublishID(scope.Context(), event.ETDeployStart, ref, deployPayload)
```

Comment:
when making an API call, this scope's context is matched to the length of the request, which means it's likely done by this point. You'll have to use `scope.AppContext()` instead, possibly wrapping `scope.AppContext()` with the current user profile first.

This applies to all uses of `scope.Context()` within this function. It'll be more maintainable to build a `ctx` variable early in the function & use that

Reply:
that's v helpful, thank you!